kinesis/src/main/scala/akka/stream/alpakka/kinesis/impl/KinesisFlowStage.scala

```diff
@@ -74,41 +63,32 @@ private[kinesis] final class KinesisFlowStage[T](
 
       private def putRecords(job: Job[T]): Future[immutable.Seq[(PutRecordsResultEntry, T)]] = {
 
-        val p = Promise[immutable.Seq[(PutRecordsResultEntry, T)]]
-
-        val request = new PutRecordsRequest()
-          .withStreamName(streamName)
-          .withRecords(job.records.map(_._1).asJavaCollection)
-
-        val handler = new AsyncHandler[PutRecordsRequest, PutRecordsResult] {
-          override def onError(exception: Exception): Unit =
-            p.failure(FailurePublishingRecords(exception))
-
-          override def onSuccess(request: PutRecordsRequest, result: PutRecordsResult): Unit = {
-            val correlatedRequestResult = result.getRecords.asScala.zip(job.records).toList
-            if (result.getFailedRecordCount > 0) {
-              val result = Result(job.attempt,
-                                  correlatedRequestResult
-                                    .filter {
-                                      case (res, _) => res.getErrorCode != null
-                                    })
-              if (job.attempt > maxRetries) failAfterResendsCallback.invoke(result)
-              else resendCallback.invoke(result)
-            } else {
-              putRecordsSuccessfulCallback.invoke(NotUsed)
-            }
-            p.success(
-              correlatedRequestResult
-                .filter {
-                  case (res, _) => res.getErrorCode == null
-                }
-                .map { case (res, (_, ctx)) => (res, ctx) }
-            )
+        val request = PutRecordsRequest
+          .builder()
+          .streamName(streamName)
+          .records(job.records.map(_._1).asJavaCollection)
+          .build()
+
+        def handleResult(result: PutRecordsResponse): immutable.Seq[(PutRecordsResultEntry, T)] = {
+          val (correlatedRequestResultSuccess, correlatedRequestResultFailure) =
+            result.records.asScala.zip(job.records).toList.partition { case (res, _) => res.errorCode eq null }
+          if (result.failedRecordCount > 0) {
```

Comment:
Next question is if we should leave it to the user to decide to fail the stream on error. Just emitting the result even if it contains errors?
That's what Alpakka Elasticsearch does https://doc.akka.io/docs/alpakka/current/elasticsearch.html#passing-data-through-elasticsearchflow

Reply:
I think it would give more control to users and reduce complexity in the library, so it's a win-win imo. I'll give that a go and remove the partial error tests.